server/types/interval.go

```diff
@@ -0,0 +1,268 @@
+	}
+	bc, _, err := b.Convert(v2)
+	if err != nil {
+		return 0, err
+	}
+
+	ab := ac.(duration.Duration)
+	bb := bc.(duration.Duration)
+	return ab.Compare(bb), nil
+}
+
+// Convert implements the DoltgresType interface.
+func (b IntervalType) Convert(val any) (any, sql.ConvertInRange, error) {
+	switch val := val.(type) {
+	case duration.Duration:
+		return val, sql.InRange, nil
+	case nil:
+		return nil, sql.InRange, nil
+	default:
+		return nil, sql.OutOfRange, fmt.Errorf("%s: unhandled type: %T", b.String(), val)
+	}
+}
+
+// Equals implements the DoltgresType interface.
+func (b IntervalType) Equals(otherType sql.Type) bool {
+	if otherExtendedType, ok := otherType.(types.ExtendedType); ok {
+		return bytes.Equal(MustSerializeType(b), MustSerializeType(otherExtendedType))
+	}
+	return false
+}
+
+// FormatValue implements the DoltgresType interface.
+func (b IntervalType) FormatValue(val any) (string, error) {
+	if val == nil {
+		return "", nil
+	}
+	return b.IoOutput(sql.NewEmptyContext(), val)
+}
+
+// GetSerializationID implements the DoltgresType interface.
+func (b IntervalType) GetSerializationID() SerializationID {
+	return SerializationID_Interval
+}
+
+// IoInput implements the DoltgresType interface.
+func (b IntervalType) IoInput(ctx *sql.Context, input string) (any, error) {
+	dInterval, err := tree.ParseDInterval(input)
+	if err != nil {
+		return nil, err
+	}
+	return dInterval.Duration, nil
+}
+
+// IoOutput implements the DoltgresType interface.
+func (b IntervalType) IoOutput(ctx *sql.Context, output any) (string, error) {
+	converted, _, err := b.Convert(output)
+	if err != nil {
+		return "", err
+	}
+	// TODO: depends on `intervalStyle` configuration variable. Defaults to `postgres`.
+	d := converted.(duration.Duration)
+	return d.String(), nil
+}
+
+// IsPreferredType implements the DoltgresType interface.
+func (b IntervalType) IsPreferredType() bool {
+	return true
+}
+
+// IsUnbounded implements the DoltgresType interface.
+func (b IntervalType) IsUnbounded() bool {
+	return false
+}
+
+// MaxSerializedWidth implements the DoltgresType interface.
+func (b IntervalType) MaxSerializedWidth() types.ExtendedTypeSerializedWidth {
+	return types.ExtendedTypeSerializedWidth_64K
+}
+
+// MaxTextResponseByteLength implements the DoltgresType interface.
+func (b IntervalType) MaxTextResponseByteLength(ctx *sql.Context) uint32 {
+	return 64
+}
+
+// OID implements the DoltgresType interface.
+func (b IntervalType) OID() uint32 {
+	return uint32(oid.T_interval)
+}
+
+// Promote implements the DoltgresType interface.
+func (b IntervalType) Promote() sql.Type {
+	return Interval
+}
+
+// SerializedCompare implements the DoltgresType interface.
+func (b IntervalType) SerializedCompare(v1 []byte, v2 []byte) (int, error) {
+	if len(v1) == 0 && len(v2) == 0 {
+		return 0, nil
+	} else if len(v1) > 0 && len(v2) == 0 {
+		return 1, nil
+	} else if len(v1) == 0 && len(v2) > 0 {
+		return -1, nil
+	}
+
+	return bytes.Compare(v1, v2), nil
```

Comment:
I don't think this works because of signed values on the leading int64, you should add a test.

Reply:
`utils.Writer` converts integers so that they're binary comparable. Part of its purpose is to allow for byte comparison when arrays/slices aren't serialized.

Reply:
Here are the tests:
https://github.com/dolthub/doltgresql/blob/d6c685e21d4f2a83a2e6eee5d697d9cff60ecb45/utils/writer_test.go#L164